File: app/src/main/java/com/breakoutplus/game/GameMode.kt
package com.breakoutplus.game

/**
 * Game mode configurations controlling lives, timers, and special rules.
 */
enum class GameMode(
    val displayName: String,
    val description: String,
    val meta: String,
    val baseLives: Int,
    val timeLimitSeconds: Int,
    val endless: Boolean,
    val godMode: Boolean,
    val rush: Boolean,
    val invaders: Boolean,
    val launchSpeed: Float
) {
    CLASSIC(
        displayName = "Classic",
        description = "Standard breakout with escalating levels and powerups.",
        meta = "Lives 3 • No timer • Balanced",
        baseLives = 3,
        timeLimitSeconds = 0,
        endless = false,
        godMode = false,
        rush = false,
        invaders = false,
        launchSpeed = 90f  // 5% reduction for better control
    ),
    TIMED(
        displayName = "Timed Challenge",
        description = "Race the clock. Clear as many bricks as possible.",
        meta = "Lives 2 • 2:30 timer • Fast",
        baseLives = 2,
        timeLimitSeconds = 150,
        endless = false,
        godMode = false,
        rush = false,
        invaders = false,
        launchSpeed = 108f  // Fast for timed challenge
    ),
    ENDLESS(
        displayName = "Endless",
        description = "Infinite levels with scaling speed and brick density.",
        meta = "Lives 3 • No timer • Scaling",
        baseLives = 3,
        timeLimitSeconds = 0,
        endless = true,
        godMode = false,
        rush = false,
        invaders = false,
        launchSpeed = 100f  // Balanced for endless
    ),
    GOD(
        displayName = "God Mode",
        description = "Practice mode. No life loss, perfect for experimentation.",
        meta = "Infinite lives • No timer",
        baseLives = 99,
        timeLimitSeconds = 0,
        endless = false,
        godMode = true,
        rush = false,
        invaders = false,
        launchSpeed = 85f  // Gentle for practice
    ),
    RUSH(
        displayName = "Level Rush",
        description = "Beat each stage before the timer expires.",
        meta = "Lives 1 • 0:45 per level • Hardcore",
        baseLives = 1,
        timeLimitSeconds = 45,
        endless = false,
        godMode = false,
        rush = true,
        invaders = false,
        launchSpeed = 115f  // Very fast for rush mode
    ),
    INVADERS(
        displayName = "Invaders",
        description = "Breakout meets space invaders. Bounce shots to clear ships while dodging fire.",
        meta = "Shielded paddle • Enemy fire • No timer",
        baseLives = 3,
        timeLimitSeconds = 0,
        endless = false,
        godMode = false,
        rush = false,
        invaders = true,
        launchSpeed = 92f
    );
}

File: ios/BreakoutPlus/BreakoutPlus/Models/GameMode.swift
//
//  GameMode.swift
//  BreakoutPlus
//
//  Game mode configurations controlling lives, timers, and special rules.
//  Direct port from Android GameMode.kt
//

import Foundation

enum GameMode: String, CaseIterable, Identifiable, Codable {
    case classic, timed, endless, god, rush, volley, tunnel, survival, invaders, zen

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .classic: return "Classic"
        case .timed: return "Timed Challenge"
        case .endless: return "Endless"
        case .god: return "God Mode"
        case .rush: return "Level Rush"
        case .volley: return "Volley"
        case .tunnel: return "Tunnel Siege"
        case .survival: return "Survival"
        case .invaders: return "Invaders"
        case .zen: return "Zen Mode"
        }
    }

    var description: String {
        switch self {
        case .classic: return "Standard breakout with escalating levels and powerups."
        case .timed: return "Race the clock. Clear as many bricks as possible."
        case .endless: return "Infinite levels with scaling speed and brick density."
        case .god: return "Practice mode. No life loss, perfect for experimentation."
        case .rush: return "Beat each stage before the timer expires."
        case .volley: return "Aim once, launch a chain of balls, then brace for descending rows."
        case .tunnel: return "Break through a fortified ring and route shots through a narrow entry tunnel."
        case .survival: return "One life. Speed ramps faster as you climb."
        case .invaders: return "Breakout meets space invaders. Bounce shots to clear ships while dodging fire."
        case .zen: return "Relaxing brick-breaking without pressure. No scores or lives to worry about."
        }
    }

    var meta: String {
        switch self {
        case .classic: return "Lives 3 • No timer • Balanced"
        case .timed: return "Lives 2 • 2:30 timer • Fast"
        case .endless: return "Lives 3 • No timer • Scaling"
        case .god: return "Infinite lives • Endless levels"
        case .rush: return "Lives 1 • 0:55 per level • Hardcore"
        case .volley: return "Lives 1 • Turn-based • Chain shots"
        case .tunnel: return "Lives 2 • Precision routing • Fortress"
        case .survival: return "Lives 1 • No timer • High speed"
        case .invaders: return "Shielded paddle • Enemy fire • No timer"
        case .zen: return "No scores • No lives • Relaxed"
        }
    }

    var baseLives: Int {
        switch self {
        case .classic, .endless, .invaders: return 3
        case .timed: return 2
        case .god, .zen: return 99
        case .rush, .volley, .survival: return 1
        case .tunnel: return 2
        }
    }

    var timeLimitSeconds: Int {
        switch self {
        case .timed: return 150  // 2:30
        case .rush: return 55    // 0:55 per level
        case .volley: return 0   // No timer
        default: return 0        // No timer
        }
    }

    var endless: Bool {
        switch self {
        case .endless, .survival, .god: return true
        default: return false
        }
    }

    var godMode: Bool {
        switch self {
        case .god, .zen: return true
        default: return false
        }
    }

    var rush: Bool {
        switch self {
        case .rush: return true
        default: return false
        }
    }

    var invaders: Bool {
        switch self {
        case .invaders: return true
        default: return false
        }
    }

    var launchSpeed: Float {
        switch self {
        case .classic: return 76.5
        case .timed: return 91.8
        case .endless: return 85.0
        case .god: return 72.25
        case .rush: return 90.0
        case .volley: return 82.0
        case .tunnel: return 79.5
        case .survival: return 92.5
        case .invaders: return 78.2
        case .zen: return 72.25
        }
    }
}
